### ISSUE TITLE
weird font display - Full-width characters overlapping

### ISSUE THREAD

**Issue by kmyk** [onivim/oni2#1475]
Full-width characters are rendered with overlapping.

![Screenshot from 2020-03-16 20-32-00](https://user-images.githubusercontent.com/2203128/76754424-73f6e500-67c5-11ea-8fd0-f20f09ea863a.png)

environment: Ubuntu 18.04.4 LTS (Bionic Beaver)

This is the same problem to as described in https://github.com/onivim/oni2/issues/1293#issuecomment-599499581

---

- I tested with Noto Sans, IPA Gothic, and VL Gothic. All fonts have the same problem.
 To test, download [NotoSansCJKjp-hinted.zip](https://noto-website-2.storage.googleapis.com/pkgs/NotoSansCJKjp-hinted.zip) from https://www.google.com/get/noto/help/cjk/ and set `"editor.fontFamily": "/path/to/NotoSansMonoCJKjp-Regular.otf"`. (Setting `"editor.config": "Noto Sans Mono CJK JP"` or `"editor.fontfamily": "/usr/share/fonts/opentype/noto/notosanscjk-regular.ttc"` installed via `apt install fonts-noto-cjk` doesn't work.)
- Test file: [tofu.md](https://github.com/onivim/oni2/files/4337959/tofu.md.txt)
- Setting `"editor.fontSmoothing": "subpixel-antialiased"` (also `"editor.fontSmoothing": "antialiased"` and `"editor.fontSmoothing": "none"`) doesn't solve this. (suggested at https://github.com/onivim/oni2/issues/1293#issuecomment-599505168)

**Comment by glennsl** [comment]
Thanks for reporting, and for the detailed information. I'll try to reproduce this myself when I get the opportunity, as I use Linux myself. Would be great if you could try to reproduce on Windows as well though, to see whether this is Linux-specific.

Could you also post the log from running `oni2 -f --trace --log-filter Window,Font`?

**Comment by kmyk** [comment]
Here is the log on Ubuntu 18.04.4 LTS.
(For Windows, please wait several days...)

```
$ oni2 -f --trace --log-filter Window,Font       
[DEBUG]   +50ms Oni2.Exception : Recording backtraces
[DEBUG]  +121ms Revery.Core.Window : Starting window creation...
[INFO]     +0ms Revery.Core.Window : Creating window Oni2 width: 800 height: 600
[INFO]   +124ms Revery.Core.Window : Window created successfully.
[DEBUG]    +0ms Revery.Core.Window : Window id: 2
[DEBUG]    +0ms Revery.Core.Window : Setting window context
[DEBUG]    +3ms Revery.Core.Window : GL setup. Checking GL version...
[DEBUG]    +0ms Revery.Core.Window : Checking GL vendor...
[DEBUG]    +0ms Revery.Core.Window : Checking GL shading language version...
[INFO]     +0ms Revery.Core.Window : OpenGL hardware info:
[INFO]     +0ms Revery.Core.Window :   renderer: Mesa DRI Intel(R) HD Graphics 520 (Skylake GT2) 
[INFO]     +0ms Revery.Core.Window :   version: 3.0 Mesa 19.2.8
[INFO]     +0ms Revery.Core.Window :   vendor: Intel Open Source Technology Center
[INFO]     +0ms Revery.Core.Window :   shadingLanguageVersion: 1.30
[DEBUG]    +0ms Revery.Core.Window : Loading icon from: /tmp/.mount_oni21D539E/usr/bin/logo.png
[DEBUG]    +1ms Revery.Core.Window : Icon loaded successfully.
[DEBUG]    +2ms Revery.Core.Window : Icon set successfully.
[DEBUG]    +0ms Revery.Core.Window : Getting window metrics
[TRACE]    +0ms Revery.Core.Window : _getScaleFactor - Linux - inferring from DPI: 1.000000
[DEBUG]    +0ms Revery.Core.Window : Metrics: DevicePixelRatio: 1.000000 ScaleFactor: 1.000000 Zoom: 1.000000 Raw Dimensions: 800x600px Framebuffer: 800x600px
[TRACE]    +0ms Revery.Core.Window : setScaledSize - calling with: 800x600
[TRACE]    +0ms Revery.Core.Window : setRawSize - calling with: 800x600
[INFO]     +0ms Revery.Core.Window : Using vsync: Immediate
[TRACE]    +4ms Revery.Core.Window : _getScaleFactor - Linux - inferring from DPI: 1.000000
[TRACE]    +0ms Revery.Core.Window : _updateMetrics - new metrics: DevicePixelRatio: 1.000000 ScaleFactor: 1.000000 Zoom: 1.000000 Raw Dimensions: 800x600px Framebuffer: 800x600px
[INFO]    +35ms Oni2.Service.Font : Loading font: FiraCode-Regular.ttf 14.000000 1
[INFO]     +0ms Oni2.Service.Font : Loading font: FiraCode-Regular.ttf 12.000000 1
[TRACE]    +0ms Oni2.Service.FontLoader : loadAndValidateEditorFont path: /tmp/.mount_oni21D539E/usr/bin/FiraCode-Regular.ttf | size: 12.000000
[TRACE]   +65ms Oni2.Service.FontLoader : loadAndValidateEditorFont path: /tmp/.mount_oni21D539E/usr/bin/FiraCode-Regular.ttf | size: 14.000000
[INFO]     +5ms Revery.FontCache : Loaded : /tmp/.mount_oni21D539E/usr/bin/FiraCode-Regular.ttf
[DEBUG]   +12ms Oni2.Service.FontLoader : Measured width: 8.000000 
[DEBUG]    +0ms Oni2.Service.FontLoader : Line height: 16.799999 
[INFO]     +2ms Revery.FontCache : Loaded : /tmp/.mount_oni21D539E/usr/bin/FiraCode-Regular.ttf
[DEBUG]    +2ms Oni2.Service.FontLoader : Measured width: 7.000000 
[DEBUG]    +0ms Oni2.Service.FontLoader : Line height: 14.400000 
[INFO]    +67ms Oni2.Service.Font : Loading font: /usr/local/share/fonts/opentype/noto/NotoSansMonoCJKjp-Regular.otf 14.000000 2
[DEBUG]    +0ms Oni2.Service.Font : Discovering font: /usr/local/share/fonts/opentype/noto/NotoSansMonoCJKjp-Regular.otf
[TRACE]    +0ms Oni2.Service.FontLoader : loadAndValidateEditorFont path: /usr/local/share/fonts/opentype/noto/NotoSansMonoCJKjp-Regular.otf | size: 14.000000
[INFO]    +11ms Revery.FontCache : Loaded : /usr/local/share/fonts/opentype/noto/NotoSansMonoCJKjp-Regular.otf
[DEBUG]    +1ms Oni2.Service.FontLoader : Measured width: 7.000000 
[DEBUG]    +0ms Oni2.Service.FontLoader : Line height: 20.271999 
[TRACE]   +36ms Revery.Core.Window : _getScaleFactor - Linux - inferring from DPI: 1.000000
[TRACE]    +0ms Revery.Core.Window : _updateMetrics - new metrics: DevicePixelRatio: 1.000000 ScaleFactor: 1.000000 Zoom: 1.000000 Raw Dimensions: 800x600px Framebuffer: 800x600px
[INFO]    +78ms Revery.FontCache : Loaded : FontAwesome5FreeSolid.otf
[INFO]    +15ms Revery.FontCache : Loaded : Inter-UI-Medium.ttf
[INFO]    +15ms Revery.FontCache : Loaded : Inter-UI-Regular.ttf
[INFO]     +7ms Revery.FontCache : Loaded : FiraCode-Regular.ttf
[INFO]     +8ms Revery.FontCache : Loaded : seti.ttf
[INFO]    +11ms Revery.FontCache : Loaded : Inter-UI-Italic.ttf
```

**Comment by glennsl** [comment]
Alright, I can definitely reproduce, and I doubt this is OS-specific. There's also several issues at play here, which make it a bit difficult:

1. The most obvious issue from the screenshot is the overlapping. I noticed there are composite graphemes here, and wonder if that is the cause of it. Unfortunately since I don't understand Japanese and can't easily identify a composite grapheme and its parts it would be helpful to have smaller test cases that demonstrate the problem and also show which individual code points are part of the involved composite graphemes. Troubleshooting this is unfortunately made even more difficult by the second issue however.

2. The cursor position is calculated incorrectly, and completely bonkers in some cases. For example, on the first line, starting on the first character and moving right works fine until the last character, where it jumps back to the start of the line. And what's even stranger is that in insert mode it behaves the same until the last character where it goes back to the beginning of the line, but then when moving right again, _past_ the last character, the cursor gets positioned _before_ it. Also, on the third line, it's easy to see that the first character has a width of two ascii characters by comparing with the second line. But moving the cursor to the next character positions it in the middle of the second glyph. suggesting the first character is measured to be three ascii characters wide.

3. When scrolling horizontally, the lines, and even parts of lines, scroll by different amounts. This may be what actually causes the overlapping, but I note it as a separate issue to make sure we check that too.

**Comment by kmyk** [comment]
It seems not OS-specific. The same problem happens on Windows 10.

<img width="601" alt="Annotation 2020-03-19 141329" src="https://user-images.githubusercontent.com/2203128/77033908-d8938900-69eb-11ea-80ad-2fd0820d40fb.png">

```
C:\Users\kimiyuki>ver

Microsoft Windows [Version 10.0.18362.720]

C:\Users\kimiyuki>oni2 -f --trace --log-filter Window,Font
[DEBUG]    +3ms Revery.Core.Window : Starting window creation...
[INFO]     +0ms Revery.Core.Window : Creating window Oni2 width: 800 height: 600
[INFO]   +332ms Revery.Core.Window : Window created successfully.
[DEBUG]    +0ms Revery.Core.Window : Window id: 1
[DEBUG]    +0ms Revery.Core.Window : Setting window context
[DEBUG]    +4ms Revery.Core.Window : GL setup. Checking GL version...
[DEBUG]    +0ms Revery.Core.Window : Checking GL vendor...
[DEBUG]    +0ms Revery.Core.Window : Checking GL shading language version...
[INFO]     +1ms Revery.Core.Window : OpenGL hardware info:
[INFO]     +0ms Revery.Core.Window :   renderer: GeForce RTX 2070 SUPER/PCIe/SSE2
[INFO]     +0ms Revery.Core.Window :   version: 4.6.0 NVIDIA 436.48
[INFO]     +0ms Revery.Core.Window :   vendor: NVIDIA Corporation
[INFO]     +0ms Revery.Core.Window :   shadingLanguageVersion: 4.60 NVIDIA
[DEBUG]    +0ms Revery.Core.Window : Loading icon from: C:\Program Files\Onivim2\logo.png
[DEBUG]    +1ms Revery.Core.Window : Icon loaded successfully.
[DEBUG]    +1ms Revery.Core.Window : Icon set successfully.
[DEBUG]    +0ms Revery.Core.Window : Getting window metrics
[TRACE]    +0ms Revery.Core.Window : _getScaleFactor - from getWin32ScaleFactor: 1.500000
[DEBUG]    +0ms Revery.Core.Window : Metrics: DevicePixelRatio: 1.000000 ScaleFactor: 1.500000 Zoom: 1.000000 Raw Dimensions: 800x600px Framebuffer: 800x600px
[TRACE]    +0ms Revery.Core.Window : setScaledSize - calling with: 800x600
[TRACE]    +0ms Revery.Core.Window : setRawSize - calling with: 1200x900
[TRACE]    +0ms Revery.Core.Window : setRawSize - calling Sdl2.Window.setSize
[TRACE]    +2ms Revery.Core.Window : setRawSize: SDL size reported after resize: 1200x900
[INFO]     +0ms Revery.Core.Window : Using vsync: Immediate
[TRACE]    +1ms Revery.Core.Window : _getScaleFactor - from getWin32ScaleFactor: 1.500000
[TRACE]    +0ms Revery.Core.Window : _updateMetrics - new metrics: DevicePixelRatio: 1.000000 ScaleFactor: 1.500000 Zoom: 1.000000 Raw Dimensions: 1200x900px Framebuffer: 1200x900px
[INFO]    +36ms Oni2.Service.Font : Loading font: FiraCode-Regular.ttf 14.000000 1
[INFO]     +0ms Oni2.Service.Font : Loading font: FiraCode-Regular.ttf 12.000000 1
[TRACE]    +0ms Oni2.Service.FontLoader : loadAndValidateEditorFont path: C:\Program Files\Onivim2\FiraCode-Regular.ttf | size: 12.000000
[INFO]     +2ms Revery.FontCache : Loaded : C:\Program Files\Onivim2\FiraCode-Regular.ttf
[DEBUG]    +1ms Oni2.Service.FontLoader : Measured width: 7.200000
[DEBUG]    +0ms Oni2.Service.FontLoader : Line height: 14.004000
[TRACE]    +0ms Oni2.Service.FontLoader : loadAndValidateEditorFont path: C:\Program Files\Onivim2\FiraCode-Regular.ttf | size: 14.000000
[DEBUG]    +1ms Oni2.Service.FontLoader : Measured width: 8.400000
[DEBUG]    +0ms Oni2.Service.FontLoader : Line height: 17.010000
[INFO]     +5ms Oni2.Service.Font : Loading font: NotoSansMonoCJKjp-Regular.otf 14.000000 2
[DEBUG]    +1ms Oni2.Service.Font : Discovering font: NotoSansMonoCJKjp-Regular.otf
WARNING: COM initialization failed.[TRACE]   +38ms Revery.Core.Window : _getScaleFactor - from getWin32ScaleFactor: 1.500000
[TRACE]    +1ms Revery.Core.Window : _updateMetrics - new metrics: DevicePixelRatio: 1.000000 ScaleFactor: 1.500000 Zoom: 1.000000 Raw Dimensions: 1200x900px Framebuffer: 1200x900px
[INFO]    +11ms Revery.FontCache : Loaded : FontAwesome5FreeSolid.otf
[INFO]     +1ms Revery.FontCache : Loaded : Inter-UI-Medium.ttf
[INFO]     +2ms Revery.FontCache : Loaded : Inter-UI-Regular.ttf
[INFO]     +6ms Revery.FontCache : Loaded : FiraCode-Regular.ttf
[INFO]     +2ms Revery.FontCache : Loaded : seti.ttf
[INFO]     +2ms Revery.FontCache : Loaded : Inter-UI-Italic.ttf
[DEBUG]  +389ms Oni2.Service.Font :   at path: C:\USERS\KIMIYUKI\APPDATA\LOCAL\MICROSOFT\WINDOWS\FONTS\NOTOSANSMONOCJKJP-REGULAR.OTF
[TRACE]    +1ms Oni2.Service.FontLoader : loadAndValidateEditorFont path: C:\USERS\KIMIYUKI\APPDATA\LOCAL\MICROSOFT\WINDOWS\FONTS\NOTOSANSMONOCJKJP-REGULAR.OTF | size: 14.000000
[INFO]    +10ms Revery.FontCache : Loaded : C:\USERS\KIMIYUKI\APPDATA\LOCAL\MICROSOFT\WINDOWS\FONTS\NOTOSANSMONOCJKJP-REGULAR.OTF
[DEBUG]    +1ms Oni2.Service.FontLoader : Measured width: 7.000000
[DEBUG]    +0ms Oni2.Service.FontLoader : Line height: 20.006001
```

**Comment by pkariz** [comment]
i'm the author of #1364 and can confirm that with the new version 2.0.3.0 (commit [a312e1e2](https://github.com/onivim/oni2/commit/a312e1e2136a9ad916d2f6d06a31a63a420ac291) the problem still exists

**Comment by CrossR** [comment]
Is this with the fix from the bottom of this PR (https://github.com/onivim/oni2/pull/1372) applied in your config? That seems to have helped a lot of people who have had the issue you are talking about @pkariz.

**Comment by pkariz** [comment]
ah i overlooked that setting. With `"editor.fontSmoothing": "subpixel-antialiased"` it works, sorry and thanks!

**Comment by PatrickMassot** [comment]
Setting `"editor.fontSmoothing": "subpixel-antialiased"` helps a bit, but you still get very weird behavior. For instance open quotes after typings accents gives:
![oni_bug](https://user-images.githubusercontent.com/14060883/78587314-0dea0480-783d-11ea-8e55-53216735f782.gif)

I love this project but I'm pretty baffled by the fact I just got an email proudly announcing version 0.4.0 when you can't type accents at all without going to GitHub to find a very partial fix...

**Comment by CrossR** [comment]
Rendering is unfortunately a complex issue, with different options working great on some platforms, and then not working at all on others, so a blanket fix becomes difficult.

I have noticed that a lot of linux users do require `subpixel-antialiased` and I've been linking then to the PR mentioned above, but the number of people I've had to point that way probably implies that we need to invest some more time into seeing if there is a smarter fix that can be applied, to decide when that option should/should not be used. @bryphe, I'm not sure if you've got any smart ideas around this awkward issue, but it is something I've seen quite a lot.

Both the rendering issue and the character width issue (and every other one tagged as a daily editor blocker!) will need to be fixed moving forward.

Oni2 is still alpha software, and we won't be considering it more than that whilst we still have `daily-editor-blockers` around, so these will all be fixed as we move from alpha to beta.

**Comment by bryphe** [comment]
@zbaylin - I think your work here: https://github.com/onivim/oni2/pull/2171 will address these issues. Would you mind looking at this in the context of #2171 ?

**Comment by zbaylin** [comment]
@bryphe sure thing!

**Comment by bryphe** [comment]
The issues here should now be resolved with the font-fallback work in #2171 (and the improvements in cursor positioning + scroll logic), along with #1608 (to properly handle the cursor position in the presence of UTF-8 characters). Let me know if you are still seeing any problems here 👍 

There is one additional issue that needs to be fixed, in regards to _inputting_ characters - #1720 - but this is a separate fix (impacting input instead of rendering) and will track there.
